core/capabilities/ccip/ccipsolana/executecodec.go

```diff
@@ -0,0 +1,259 @@
+package ccipsolana
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	agbinary "github.com/gagliardetto/binary"
+	"github.com/gagliardetto/solana-go"
+	ag_solanago "github.com/gagliardetto/solana-go"
+	chainsel "github.com/smartcontractkit/chain-selectors"
+
+	"github.com/smartcontractkit/chainlink-ccip/chains/solana/gobindings/ccip_router"
+	cciptypes "github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+// ExecutePluginCodecV1 is a codec for encoding and decoding execute plugin reports.
+// Compatible with:
+// - "OffRamp 1.6.0-dev"
+type ExecutePluginCodecV1 struct {
+}
+
+func NewExecutePluginCodecV1() *ExecutePluginCodecV1 {
+	return &ExecutePluginCodecV1{}
+}
+
+func (e *ExecutePluginCodecV1) Encode(ctx context.Context, report cciptypes.ExecutePluginReport) ([]byte, error) {
+	if len(report.ChainReports) == 0 || len(report.ChainReports) > 1 {
+		return nil, fmt.Errorf("unexpected chain report length: %d", len(report.ChainReports))
+	}
+
+	var message ccip_router.Any2SolanaRampMessage
+	chainReport := report.ChainReports[0]
+	if len(chainReport.Messages) > 0 {
+		// currently only allow committing one message at a time
+		msg := chainReport.Messages[0]
+		tokenAmounts := make([]ccip_router.Any2SolanaTokenTransfer, 0, len(msg.TokenAmounts))
+		for _, tokenAmount := range msg.TokenAmounts {
+			if tokenAmount.Amount.IsEmpty() {
+				return nil, fmt.Errorf("empty amount for token: %s", tokenAmount.DestTokenAddress)
+			}
+
+			if len(tokenAmount.DestTokenAddress) != solana.PublicKeyLength {
+				return nil, fmt.Errorf("invalid destTokenAddress address: %v", tokenAmount.DestTokenAddress)
+			}
+
+			tokenAmounts = append(tokenAmounts, ccip_router.Any2SolanaTokenTransfer{
+				SourcePoolAddress: tokenAmount.SourcePoolAddress,
+				DestTokenAddress:  solana.PublicKeyFromBytes(tokenAmount.DestTokenAddress),
+				ExtraData:         tokenAmount.ExtraData,
+				Amount:            bigIntToBytes32(tokenAmount.Amount),
+				DestGasAmount:     bytesToUint32(tokenAmount.DestExecData),
+			})
+		}
+
+		family, err := chainsel.GetSelectorFamily(uint64(chainReport.SourceChainSelector))
+		if err != nil {
+			return nil, fmt.Errorf("invalid source chain selector: %w", err)
+		}
+
+		// if source chain is Solana the Borsh decoder will be used, for EVM we will construct the extra args from
+		// the chain agnostic extra args map
+		var extraArgs ccip_router.SolanaExtraArgs
+		switch family {
+		case chainsel.FamilySolana:
+			decoder := agbinary.NewBorshDecoder(msg.ExtraArgs)
+			err = extraArgs.UnmarshalWithDecoder(decoder)
+			if err != nil {
+				return nil, fmt.Errorf("invalid extra arguments: %w", err)
+			}
+		case chainsel.FamilyEVM:
+			extraArgs, err = parseExtraArgsMap(msg.ExtraArgsDecoded)
+			if err != nil {
+				return nil, fmt.Errorf("invalid extra args map: %w", err)
+			}
+		default:
+			return nil, fmt.Errorf("unsupported source chain family: %s", family)
+		}
```

Comment:
Shouldn't logic like this be in the extra args codec instead of here?

Reply:
The extra args [codec](https://github.com/smartcontractkit/chainlink/pull/16016) is used for constructing the ExtraArgsDecoded map when required by the codec. From my understanding, if the map is needed based on the source chain, we can customize its handling to build the extra arguments during encoding.

Reply:
I need to construct a `ccip_router.SolanaExtraArgs` from this decoded map, so I can return the borsh encode into bytes